File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
from sys import stdin

input = stdin.readline
k, n = map(int, input().rstrip().split())
arr = [int(input()) for _ in range(k)]
end = sum(arr)//n

def line_count(mid):    # mid로 나누면 몇개의 선이 되는가?
    count = 0
    for i in arr:
        count += i//mid     # arr에 들어있는 원소마다 mid로 나누었을 때 나오는 개수를 더함
    return count

def binarySearch(start, end, n):
    if start > end:
        return end
    mid = (start + end) // 2
    count = line_count(mid)
    if count >= n:  # 선의 개수가 요구하는 것보다 많거나 같으면, 길이를 늘려서 나오는 양을 줄여야지.
        return binarySearch(mid+1, end, n)
    else:   # 선의 개수가 요구하는 것보다 적으면, 길이를 줄여서 나오는 양을 늘려야지
        return binarySearch(start, mid-1, n)

print(binarySearch(1, end, n))

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
	public static void main(String[] args) throws NumberFormatException, IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();
		StringTokenizer st = new StringTokenizer(br.readLine());
		int K = Integer.parseInt(st.nextToken());
		int N = Integer.parseInt(st.nextToken());
		int[] arr = new int[K];
		int max = 0;
		//System.out.println(K + " " + N);
		for (int i = 0; i < K; i++) {
			int tmp = Integer.parseInt(br.readLine());
			//System.out.println(tmp);
			arr[i] = tmp;
			max = tmp > max ? tmp : max;
		}
		
		System.out.println(binarySearch(arr, N, max));
		
	}
	
	public static int binarySearch(int[] list, int n, int end) {
		int start = 0;
		
		while (start < end) {
			int mid = (int) ((start + (long) end + 1) / 2);
			
			/*
			 * mid를 기준으로 잘랐을 때, 랜선의 개수가 n개 이상이라면
			 * start를 mid로 두고, 아니라면 end를 mid - 1로 둔다.
			 */
			
			int line = 0;
			for (int i = 0; i < list.length; i++) {
				line += list[i] / mid;
			}
			//System.out.println("line : " + line);
			if (line >= n) {
				start = mid;
			} else {
				end = mid - 1;
			}
		}
		
		return start;
	}
}